packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts

```diff
@@ -0,0 +1,829 @@
+    );
+  }
+
+  private init() {
+    this.current = 0;
+    this.errors = [];
+    this.invalid = [];
+    this.contextStack = new ParsingContextStack();
+  }
+
+  private advance(): SyntaxToken {
+    return this.tokens[this.current++];
+  }
+
+  private peek(lookahead: number = 0): SyntaxToken | undefined {
+    if (lookahead + this.current >= this.tokens.length) {
+      return undefined;
+    }
+
+    return this.tokens[this.current + lookahead];
+  }
+
+  private match(...kind: SyntaxTokenKind[]): boolean {
+    const checkRes = this.check(...kind);
+    if (checkRes) {
+      this.advance();
+    }
+
+    return checkRes;
+  }
+
+  private check(...kind: SyntaxTokenKind[]): boolean {
+    const currentToken = this.peek();
+    if (!currentToken) {
+      return false;
+    }
+
+    return kind.includes(currentToken.kind);
+  }
+
+  private previous(): SyntaxToken {
+    return this.tokens[this.current - 1];
+  }
+
+  private consume(message: string, ...kind: SyntaxTokenKind[]) {
+    if (!this.match(...kind)) {
+      const invalidToken = this.peek()!;
+      const error = new ParsingError(
+        ParsingErrorCode.EXPECTED_THINGS,
+        message,
+        invalidToken.offset,
+        invalidToken.offset + invalidToken.length,
+        invalidToken,
+      );
+      this.invalid.push(invalidToken);
+      this.errors.push(error);
+      throw error;
+    }
+  }
+
+  parse(): Result<SyntaxNode & { kind: SyntaxNodeKind.PROGRAM }, ParsingError> {
+    const body: ElementDeclarationNode[] = [];
+
+    this.init();
+
+    while (this.peek() && this.peek()?.kind !== SyntaxTokenKind.EOF) {
+      try {
+        body.push(this.elementDeclaration());
+      } catch (e) {
+        if (!(e instanceof ParsingError)) {
+          throw e;
+        }
+        if (!(this.peek()?.kind === SyntaxTokenKind.EOF)) {
+          this.invalid.push(this.advance());
+        } else {
+          const eof = this.peek()!;
+          this.invalid.push(eof);
+          this.errors.push(
+            this.generateTokenError(eof, ParsingErrorCode.INVALID, 'Unexpected EOF'),
+          );
+        }
+      }
+    }
+
+    const eof = this.advance();
+    const program = new ProgramNode({ body, eof, invalid: this.invalid });
+
+    return new Result(program, this.errors);
+  }
+
+  private elementDeclaration(): ElementDeclarationNode {
+    this.consume('Expect identifier', SyntaxTokenKind.IDENTIFIER);
+    const type = this.previous();
+    let name: NormalFormExpressionNode | undefined;
+    let as: SyntaxToken | undefined;
+    let alias: NormalFormExpressionNode | undefined;
+
+    if (
+      this.peek()?.kind !== SyntaxTokenKind.COLON &&
+      this.peek()?.kind !== SyntaxTokenKind.LBRACE
+    ) {
+      try {
+        name = this.normalFormExpression();
+      } catch (e) {
+        if (e instanceof ParsingError) {
+          this.synchronizeElementDeclarationName();
+        } else {
+          throw e;
+        }
```

Comment:
This logic is duplicated in multiple places, suggest we should have a dedicated function for this